primitives/version/proc-macro/src/lib.rs

```diff
@@ -0,0 +1,68 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An attribute that accepts a version declaration of a runtime and generates a custom wasm section
+//! with the equivalent contents.
+//!
+//! The custom section allows ti read the version of the runtime without having to execute any code.
+//! Instead, the generated custom section can be relatively easily parsed from the wasm binary. The
+//! identifier of the custom section is "runtime_version".
+//!
+//! A shortcoming of this macro is that it is unable to embed information regarding supported APIs.
+//! This is supported by the `construct_runtime!` macro.
+//!
+//! This macro accepts a const item like the following:
+//!
+//! ```rust,ignore
+//! #[sp_version_proc_macro::runtime_version]
+//! pub const VERSION: RuntimeVersion = RuntimeVersion {
+//! 	spec_name: create_runtime_str!("test"),
+//! 	impl_name: create_runtime_str!("test"),
+//! 	authoring_version: 10,
+//! 	spec_version: 265,
+//! 	impl_version: 1,
+//! 	apis: RUNTIME_API_VERSIONS,
+//! 	transaction_version: 2,
+//! };
+//! ```
+//!
+//! and will pass it through and add code required for emit a custom section. The information that
+//! will go into the custom section is parsed from the item declaration. Due to that, the macro is
+//! somewhat rigid in terms of the code it accepts. There are the following considerations:
+//!
+//! - The `spec_name` and `impl_name` must be set by a macro-like expression. The name of the macro
```

Comment:
Funnily, this is similar to my initial approach. I found it confusing that the actual type is `RuntimeString`, but the user passes an `&str` expression. That's why I ditched it.

This way where we try to recognize the a macro call is very rigid, but at least it seems less magicky.

Reply:
What is magicky in taking a &str?

Reply:
In Rust types are not converted under the hood typically. There are certain exceptions like `Deref` coercions, but in overwhelming majority of cases you have to be explicit if you want conversion to happen. You cannot even use a string literal to initialize a `String`!

Applying this logic to the interface in question, if we were to accept a string literal directly and wrap it into `RuntimeString::Borrowed` then it would look like the string type was transmogrified into `RuntimeString`. 

It may have been OK in other cases, but here we have syntax which is identical to normal Rust. Imagine a reader who is not familiar with the internal workings of the macro. If the macro there, then suddenly rules are different and a string literal can be accepted in place of a `RuntimeString` defying the standard rules. One may call such a behavior "magical".